src/CoreWCF.Primitives/src/CoreWCF/Dispatcher/InvokerUtil.cs

```diff
@@ -104,54 +106,123 @@ internal static InvokeDelegate GenerateInvokeDelegate(MethodInfo method, out int
-                    try
-                    {
-                        if (returnsValue)
-                        {
-                            result = method.Invoke(target, paramsLocal);
-                        }
-                        else
-                        {
-                            method.Invoke(target, paramsLocal);
-                        }
-                    }
-                    catch (TargetInvocationException tie)
-                    {
-                        ExceptionDispatchInfo.Capture(tie.InnerException).Throw();
-                    }
+                // delegate (object target, object[] inputs, object[] outputs)
+                // {
+                //     object[] paramsLocal = null;
+                //     if (paramCount > 0)
+                //     {
+                //         paramsLocal = new object[paramCount];
+                //
+                //         for (int i = 0; i < inputPos.Length; i++)
+                //         {
+                //             paramsLocal[inputPos[i]] = inputs[i];
+                //         }
+                //     }
+                //
+                //     object result = null;
+                //     try
+                //     {
+                //         if (returnsValue)
+                //         {
+                //             result = method.Invoke(target, paramsLocal);
+                //         }
+                //         else
+                //         {
+                //             method.Invoke(target, paramsLocal);
+                //         }
+                //     }
+                //     catch (TargetInvocationException tie)
+                //     {
+                //         ExceptionDispatchInfo.Capture(tie.InnerException).Throw();
+                //     }
+                //
+                //     for (int i = 0; i < outputPos.Length; i++)
+                //     {
+                //         Debug.Assert(paramsLocal != null);
+                //         outputs[i] = paramsLocal[outputPos[i]];
+                //     }
+                //
+                //     return result;
+                // };
 
-                    for (int i = 0; i < outputPos.Length; i++)
-                    {
-                        Debug.Assert(paramsLocal != null);
-                        outputs[i] = paramsLocal[outputPos[i]];
-                    }
+                var targetParam = Expression.Parameter(typeof(object), "target");
+                var inputsParam = Expression.Parameter(typeof(object[]), "inputs");
+                var outputsParam = Expression.Parameter(typeof(object[]), "outputs");
+                var paramsLocal = Expression.Variable(typeof(object[]), "paramsLocal");
+                var result = Expression.Variable(typeof(object), "result");
+
+                var methodParam = Expression.Constant(method);
+
+                var assignParamsLocal = Expression.Assign(paramsLocal, Expression.Condition(
```

Comment:
You don't need to have the condition on the paramCount be part of the expression. You can conditionally emit a different expression based on the paramCount > 0. This is part of the performance improvement moving to Expression's, you can emit less code based on conditions.

Reply:
done!